standalone-metastore/metastore-server/src/main/java/org/apache/hadoop/hive/metastore/txn/TxnHandler.java

```diff
@@ -5283,6 +5284,39 @@ is performed on that db (e.g. show tables, created table, etc).
             return response;
           }
         }
+
+        if (isValidTxn(txnId)) {
+          LockType lockType = LockTypeUtil.getLockTypeFromEncoding(lockChar)
+                  .orElseThrow(() -> new MetaException("Unknown lock type: " + lockChar));
+
+          if (lockType == LockType.EXCL_WRITE && blockedBy.state == LockState.ACQUIRED) {
```

Comment:
I don't think we should give up if there is already CTAS on the same table.  What happens if the 1st CTAS aborts?

Reply:
We do not know at what stage the 1st query can abort. As it is non-deterministic, we needed to make an assumption.

So when this is enabled via the conf, we will be optimistic about the outcome and assume the 1st query always succeeds. With this assumption, we can fail-early the 2nd concurrent ctas query and prevent any unnecessary move tasks and clean up that would have been associated with the 2nd query, if it was to continue until the commit stage. Also, the 2nd user will not have to wait for a long time to find out that the query failed.

But when this feature is disabled, the query will run with a pessimistic assumption that the 1st query can abort. As a result, it does not fail the 2nd query until the commit stage. This will result in a lot of overhead and clean-up associated with the failed query.  This may also make the user wait for a long time only to find out that the query failed which I think is not ideal.

This was my thought process. Would this be fine?

Reply:
I don't really like that we are adding extra overhead in checkLocks method, it's already a sensitive part performance-wise. I think we should try to optimize: if it's CTAS we know that it could only be blocked by another artificial CTAS or DROP database, so no need to run expensive checkLock `BIG` query.
Also, I would expect IOW to behave similarly to CTAS, currently it doesn't fail and is executed in sequential order.

Reply:
I was able to optimize it by moving this check to a much earlier step, right before we call checkLocks. This prevents unnecessary insert into the TXN_COMPONENTS table and saves us at least 3 to 4 subsequent queries to the metastore.